File: src/recommender/catalogue.py
""" Functions for getting specific things from collection, eg. performers, studios, ... """
from datetime import datetime

from src.schemas import VideoData, CatalogueQuery


def get_catalogue(videos_list: list[VideoData], query: CatalogueQuery) -> dict[str, list[tuple[str, int]]]:
    """
    For a given set of filters (performer, studio, collection, tags), for each
    category return map of scene counts. 
    """
    
    q = query
    
    # apply sort performers choice
    if q.use_primary_actors:
        get_video_actors = get_video_primary_actors
    else:
        get_video_actors = get_video_actors_all

    # filter
    if q.filter_actor:        videos_list = [ vd for vd in videos_list if q.filter_actor in         get_video_actors(vd) ]
    if q.filter_studio:       videos_list = [ vd for vd in videos_list if q.filter_studio in        get_video_studios(vd) ]
    if q.filter_collection:   videos_list = [ vd for vd in videos_list if q.filter_collection in    get_video_collection(vd) ]
    if q.filter_tag:          videos_list = [ vd for vd in videos_list if q.filter_tag in           get_video_tags(vd) ]
    
    # get item counts
    result = {}
    result['actor_info'] =      _get_item_info(videos_list, get_video_actors)
    result['studio_info'] =     _get_item_info(videos_list, get_video_studios)
    result['collection_info'] = _get_item_info(videos_list, get_video_collection)
    result['tag_info'] =        _get_item_info(videos_list, get_video_tags)
    
    # sort
    for item_type, tuple_list in result.items():
        if q.query_string and item_type == q.query_type:
            # tf-idf stuff
            ...
        else:
            result[item_type] = sorted(tuple_list, reverse=True, key=lambda x: x[1])
    
    return result



def _get_item_info(videos_list: list[VideoData], selector_func) -> list[tuple]:
    """  """
    item_counts = {}
    newest_video = {}
    oldest_video = {}
    new_videos = {}
    for video_data in videos_list:
        for item in selector_func(video_data):
            item = item.lower().replace('.', '').strip()
            item_counts[item] = item_counts.get(item, 0) + 1
            newest_video[item] = max( newest_video.get(item, '0000'), video_data.date_added )
            oldest_video[item] = min( oldest_video.get(item, '2100'), video_data.date_added )
            if seconds_from_now(video_data.date_added) < (60*60*24*7): # a week old
                new_videos[item] = new_videos.get(item, 0) + 1
    return [
        (item, item_counts.get(item), newest_video.get(item), new_videos.get(item))
        for item in item_counts.keys()
    ]


def get_video_actors_all(vd: VideoData) -> list[str]:
    return vd.actors

def get_video_primary_actors(vd: VideoData) -> list[str]:
    return vd.primary_actors

def get_video_studios(vd: VideoData) -> list[str]:
    return [ x for x in [vd.studio, vd.line] if x is not None ]

def get_video_collection(vd: VideoData) -> list[str]:
    if vd.collection:
        return [vd.collection]
    return []

def get_video_tags(vd: VideoData) -> list[str]:
    return vd.tags


# HELPERS

def seconds_from_now(date: str) -> int:
    fmt = "%Y-%m-%d %H:%M"
    curr = datetime.now().strftime(fmt)
    return _dates_difference(date, curr, fmt=fmt)

def _dates_difference(d1: str, d2: str, fmt="%Y-%m-%d %H:%M") -> int:
    dt1 = datetime.strptime(d1, fmt)
    dt2 = datetime.strptime(d2, fmt)
    return int((dt2 - dt1).total_seconds())

File: go_backend/internal/query/catalogue.go
package query

import (
	"log"
	"slices"
	"strings"
	"time"

	"cpv_backend/internal/schemas"
)

type ItemInfo struct {
	Name 			string
	VideoCount 		int
	NewestVideo 	string
	NewVideoCount 	int
}

type Catalogue struct {
	ActorInfo      []ItemInfo `json:"actor_info"`
	StudioInfo     []ItemInfo `json:"studio_info"`
	CollectionInfo []ItemInfo `json:"collection_info"`
	TagInfo        []ItemInfo `json:"tag_info"`
	TimeTakenMS    float64    `json:"time_taken_ms"`
}

// #region - PUB --------------------------------------------------------------------------------------------------------

// GetCatalogue gets info on videos grouped by different item types, (eg. actor)
func GetCatalogue(vids []schemas.VideoData, q schemas.CatalogueQuery) (Catalogue, error) {

	start := time.Now()
	var cat Catalogue

	// determine actor type
	getVideoActors := getVideoActors_All
	if q.UsePrimaryActors {
		getVideoActors = getVideoActors_Primary
	}

	/* FILTER */
	if q.FilterActor != "" {
		lw := strings.ToLower(q.FilterActor)
		vids = filterSliceFunc(vids, func(vd schemas.VideoData) bool {
			return slices.Contains(getVideoActors(vd), lw)
		})
	}
	if q.FilterStudio != "" {
		lw := strings.ToLower(q.FilterStudio)
		vids = filterSliceFunc(vids, func(vd schemas.VideoData) bool {
			return slices.Contains(getVideoStudios(vd), lw)
		})
	}
	if q.FilterCollection != "" {
		lw := strings.ToLower(q.FilterCollection)
		vids = filterSliceFunc(vids, func(vd schemas.VideoData) bool {
			return slices.Contains(getVideoCollection(vd), lw)
		})
	}
	if q.FilterTag != "" {
		lw := strings.ToLower(q.FilterTag)
		vids = filterSliceFunc(vids, func(vd schemas.VideoData) bool {
			return slices.Contains(getVideoTags(vd), lw)
		})
	}
	
	
	/* GET ITEM INFOS */
	cat.ActorInfo = 	 getItemInfo(vids, getVideoActors)
	cat.StudioInfo = 	 getItemInfo(vids, getVideoStudios)
	cat.CollectionInfo = getItemInfo(vids, getVideoCollection)
	cat.TagInfo = 		 getItemInfo(vids, getVideoTags)

	
	/* TF-IDF */
	if q.QueryString != "" {
		// ...
	}
	
	cat.TimeTakenMS = float64(time.Since(start).Microseconds())/1000
	return cat, nil
	
}


// #region - PRIV ------------------------------------------------------------------------------------------------------


// 
func getItemInfo(vids []schemas.VideoData, item_extractor_func func(schemas.VideoData)[]string) []ItemInfo {

	item_counts := map[string]int{}
	newest_video := map[string]string{}
	// oldest_video := map[string]string{}
	new_videos := map[string]int{}

	for _, vd := range vids {
		for _, item := range item_extractor_func(vd) {
			item_counts[item] = item_counts[item] + 1
			// update item newest video
			newest_date_added, ok := newest_video[item]
			if !ok {
				newest_date_added = "1970-01-01T00:01"
			}
			if vd.DateAdded > newest_date_added {
				newest_video[item] = vd.DateAdded
			}
			// update new videos
			if secondsFromNow(vd.DateAdded) < (60*60*24*7) {
				new_videos[item] = new_videos[item] + 1
			}
		}
	}

	var item_infos []ItemInfo
	for k := range item_counts {
		info := ItemInfo{
			Name: 			k,
			VideoCount: 	item_counts[k],
			NewestVideo: 	newest_video[k],
			NewVideoCount: 	new_videos[k],
		}
		item_infos = append(item_infos, info)
	}
	return item_infos
}

/* ITEM GETTERS */

func getVideoActors_All(vd schemas.VideoData) []string {
	slic := []string{}
	for _, a := range vd.Actors {
		slic = append(slic, standardizeString(a))
	}
	return slic
}

func getVideoActors_Primary(vd schemas.VideoData) []string {
	slic := []string{}
	for _, a := range vd.PrimaryActors {
		slic = append(slic, standardizeString(a))
	}
	return slic
}

func getVideoStudios(vd schemas.VideoData) []string {
	var slic []string
	if vd.Studio != "" {
		slic = append(slic, standardizeString(vd.Studio))
	}
	if vd.Line != "" {
		slic = append(slic, standardizeString(vd.Line))
	}
	return slic
}

func getVideoCollection(vd schemas.VideoData) []string {
	return []string{standardizeString(vd.Collection)}
}

func getVideoTags(vd schemas.VideoData) []string {
	slic := []string{}
	for _, a := range vd.Tags {
		slic = append(slic, standardizeString(a))
	}
	return slic
}

// #region - HELPERS ---------------------------------------------------------------------------------------------------


func secondsFromNow(date string) float64 {
    const layout = "2006-01-02 15:04"
    t, err := time.Parse(layout, date)
    if err != nil {
        log.Println("error parsing date: "+err.Error())
        return 0
    }
    duration := time.Since(t)
    return duration.Seconds()
}
